executor/write.go

```diff
@@ -44,114 +44,118 @@ var (
 //     4. err (error) : error in the update.
 func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datum, modified []bool, t table.Table,
 	onDup bool) (bool, bool, int64, error) {
-	var sc = ctx.GetSessionVars().StmtCtx
-	var changed, handleChanged = false, false
+	sc := ctx.GetSessionVars().StmtCtx
+	changed, handleChanged := false, false
 	// onUpdateSpecified is for "UPDATE SET ts_field = old_value", the
 	// timestamp field is explicitly set, but not changed in fact.
-	var onUpdateSpecified = make(map[int]bool)
+	onUpdateSpecified := make(map[int]bool)
 	var newHandle int64
 
 	// We can iterate on public columns not writable columns,
 	// because all of them are sorted by their `Offset`, which
 	// causes all writable columns are after public columns.
+
+	// 1. Cast modified values.
 	for i, col := range t.Cols() {
 		if modified[i] {
 			// Cast changed fields with respective columns.
 			v, err := table.CastValue(ctx, newData[i], col.ToInfo())
 			if err != nil {
-				return false, handleChanged, newHandle, errors.Trace(err)
+				return false, false, 0, errors.Trace(err)
 			}
 			newData[i] = v
 		}
+	}
 
-		if mysql.HasNotNullFlag(col.Flag) && newData[i].IsNull() && sc.BadNullAsWarning {
-			var err error
-			newData[i], err = table.GetColDefaultValue(ctx, col.ToInfo())
-			if err != nil {
-				return false, handleChanged, newHandle, errors.Trace(err)
-			}
-		}
-		// Rebase auto increment id if the field is changed.
-		if mysql.HasAutoIncrementFlag(col.Flag) {
-			if newData[i].IsNull() {
-				return false, handleChanged, newHandle, table.ErrColumnCantNull.GenByArgs(col.Name)
-			}
-			val, errTI := newData[i].ToInt64(sc)
-			if errTI != nil {
-				return false, handleChanged, newHandle, errors.Trace(errTI)
-			}
-			err := t.RebaseAutoID(ctx, val, true)
-			if err != nil {
-				return false, handleChanged, newHandle, errors.Trace(err)
+	// 2. Check null.
+	for i, col := range t.Cols() {
+		if err := col.CheckNotNull(newData[i]); err != nil {
+			if sc.BadNullAsWarning {
```

Comment:
The old code doesn't have this logic?

Reply:
It has this at old file line 67, 76 and 109.
Line 67 is used for BadNullAsWarning only, line 76 is used for auto ID and line 109 checks all the not null constraint.